### PR TITLE
Fix build errors caused by conflicts between windows.h and antlr4-runtime.h

### PR DESCRIPTION
**Pull request by a-zakir**

### CODE REVIEW

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h [hunk 1]
```diff
@@ -269,5 +293,13 @@ class NodeVisitor: public IName
      * @return The result of processing the ComponentParameterNode.
      */
     virtual R visit(const Nodes::ComponentParameterNode*, Args... args) = 0;
+
+    void setNodeVisitorCustomLog(const NodeVisitorCustomLog& nodeVisitorCustomLog)
+    {
+        nodeVisitorCustomLog_ = nodeVisitorCustomLog;
+    }
```
Comment (guilpier-code) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: About : **setNodeVisitorCustomLog(...)**
is unused. 
Removal ?

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h (hunk 1 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: The client of this class has to use this method in order to customize logging;

For instance the client would create an  NodeVisitorCustomLog cust_log = {.info = my_info_log, .warning =etc...};
where my_info could be:

void my_info_log(const std::string& msg){
logs.info(msg);
}

File: src/solver/expressions/visitors/EvalVisitor.cpp [hunk 2]
```diff
@@ -21,6 +21,7 @@
 
 #include "antares/solver/expressions/visitors/EvalVisitor.h"
 
+#include <cmath>
```
Comment (guilpier-code) [inline src/solver/expressions/visitors/EvalVisitor.cpp:24]: `#include <cmath>` : seems useless (removing it still allows a successful compilation)

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h [hunk 3]
```diff
@@ -19,18 +19,33 @@
...
+    using LogFunction = std::function<void(const std::string&)>;
+
+    LogFunction info;
+    LogFunction warning;
+    LogFunction error;
+};
+
+void ToYuniInfo(const std::string& msg);
+
+void ToYuniError(const std::string& msg);
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: This declaration is unused

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h [hunk 4]
```diff
@@ -269,5 +284,13 @@ class NodeVisitor: public IName
...
      */
     virtual R visit(const Nodes::ComponentParameterNode*, Args... args) = 0;
+
+    void setNodeVisitorCustomLog(const NodeVisitorCustomLog& nodeVisitorCustomLog)
+    {
+        nodeVisitorCustomLog_ = nodeVisitorCustomLog;
+    }
+
+private:
+    NodeVisitorCustomLog nodeVisitorCustomLog_ = RedirectToStandardOutputs();
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]:
```suggestion
    LogSink nodeVisitorCustomLog_ = RedirectToAntaresLogs();
```

File: src/solver/expressions/visitors/NodeVisitor.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+{
+void ToYuniInfo(const std::string& msg)
+{
+    logs.info() << msg;
+}
+
+inline void ToYuniError(const std::string& msg)
+{
+    logs.error() << msg;
+}
```
Comment (flomnes) [inline src/solver/expressions/visitors/NodeVisitor.cpp]:
```suggestion
static void ToYuniInfo(const std::string& msg)
{
    logs.info() << msg;
}

static void ToYuniError(const std::string& msg)
{
    logs.error() << msg;
}
```

File: src/solver/expressions/visitors/NodeVisitor.cpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+NodeVisitorCustomLog RedirectToStandardOutputs()
+{
+    return {.info = ToYuniInfo, .warning = ToYuniInfo, .error = ToYuniError};
```
Comment (flomnes) [inline src/solver/expressions/visitors/NodeVisitor.cpp]: Please add a proper warning function

```suggestion
    return {.info = ToYuniInfo, .warning = ToYuniWarning, .error = ToYuniError};
```

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h [hunk 7]
```diff
@@ -119,7 +134,7 @@ class NodeVisitor: public IName
         }
         catch (std::exception&)
         {
-            logs.error() << "Antares::Solver::Visitor: could not visit the node!";
+            nodeVisitorCustomLog_.error("Antares::Solver::Visitor: could not visit the node!");
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]:
```suggestion
            log_.error("Antares::Solver::Visitor: could not visit the node!");
```

File: src/solver/expressions/visitors/NodeVisitor.cpp
Comment (flomnes) [inline src/solver/expressions/visitors/NodeVisitor.cpp:1]: Please rename this function to something like logSink, etc.

File: src/solver/expressions/visitors/NodeVisitor.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+{
+    logs.info() << msg;
+}
+
+inline void ToYuniError(const std::string& msg)
+{
+    logs.error() << msg;
+}
+
+NodeVisitorCustomLog RedirectToStandardOutputs()
```
Comment (flomnes) [inline src/solver/expressions/visitors/NodeVisitor.cpp]:
```suggestion
LogSink RedirectToAntaresLogs()
```

**Review by flomnes — CHANGES_REQUESTED** [review]
Thanks for this, it works

As @guilpier-code said, you can remove the setter since it's unused (for now)

Please add a comment explaining why we need a LogSink (antlr4-runtime.h & windows.h conflict)

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h [hunk 9]
```diff
@@ -19,18 +19,33 @@
 ** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
 */
 #pragma once
+#include <functional>
+#include <iostream>
 #include <optional>
 #include <typeindex>
 #include <vector>
 
-#include <antares/logs/logs.h>
```
Comment (guilpier-code) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h:26]: About `#include <antares/logs/logs.h>` : 
As we know, this is the inclusion that causes a compilation error on Windows.
Just to be clear, this header indirectly `#include <windows.h>`, that holds an incompatibility with **antlr4**.
As a proof, replacing `#include <antares/logs/logs.h>` with `#include <windows.h>` in the current **logs.h** causes the same compilation error.
Note that `#include <winsock2.h>` is also a cause for incompatibility, as this header includes **windows.h**

**Comment by a-zakir** [comment]
> Thanks for this, it works
> 
> As @guilpier-code said, you can remove the setter since it's unused (for now)
> 
> Please add a comment explaining why we need a LogSink (antlr4-runtime.h & windows.h conflict)

in fact, the method is no longer needed

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h (hunk 1 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: [update] the default log is now yuni::logs, i can safely remove the setter

File: src/solver/expressions/visitors/EvalVisitor.cpp (hunk 2 above)
Comment (a-zakir) [reply src/solver/expressions/visitors/EvalVisitor.cpp:24]: done

File: src/solver/expressions/visitors/NodeVisitor.cpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+
+#include <antares/logs/logs.h>
+#include <antares/solver/expressions/visitors/NodeVisitor.h>
+
+namespace Antares::Solver::Visitors
+{
+void ToYuniInfo(const std::string& msg)
```
Comment (guilpier-code) [inline src/solver/expressions/visitors/NodeVisitor.cpp]: We can do it like that.
Or we could define an interface class (with the same public methods as current **Yuni** logger), and create a **Yuni** implementation, internally calling the **Yuni** logger.

**Comment by guilpier-code** [comment]
We identified 3 solutions in order to solve the compilation issue, due to the current conflict.
(the first 2 are quite the same, they count as one solution).
- Do as it's solved now (replacing direct call to **Yuni** logger with functions defined in a **.cpp** that includes the **Yuni** logger header **logs.h**)
- We also can create a logger interface, as suggested [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2475#discussion_r1843458761)
- change **Yuni** code, in order to include `<mutex>` instead of system custom headers related to multi threading.
- Replace **Yuni** logger with another one, compatible with **antlr**. This new logger should replace **Yuni** logger at some point anyway.

File: src/solver/expressions/visitors/NodeVisitor.cpp (hunk 10 above)
Comment (a-zakir) [reply src/solver/expressions/visitors/NodeVisitor.cpp]: Yes, same spirit as LogSink; the interface implies using a pointer or something like that, injected into the ctor? or from a setter?

Comment (flomnes) [reply src/solver/expressions/visitors/NodeVisitor.cpp]: No need for an interface class at this point

**Review by flomnes — APPROVED** [review]

**Pull request merged**